In order to link the variant calling data to our genes a few scripts were generated in python v3.7.

The first script that was used is "get_substitution_rates.py". This program was used to divide the mutations into three groups;single nucleotide polymorphism, double mutation or long region substitution. This program takes a tab delimited variant summary fileas input. This file should contain the Chromosome ID, the position of the mutation, the original nucleotide and the altered nucleotide.The output is a txt file that contains the total number of snpâ€™s, the number of long region substitutions, the number of double mutationsand a substitution matrix. This matrix contains the total number of substitutions per nucleotide change. These changes are describedas A > T, A > C etc.The second script that was used is "all.py" or "all_blast.py".

The scripts were used to link the substitution information to thegenes and generates a tab delimited file that can be used to find the sequences of these mutated genes. Both scripts work the same butall_blast.py has been altered so it can handle the blast results and the prodigal files. The script "all.py" needs a sequence file, this is thefasta file of the metagenome, a mutation file, this is a tab delimited variant summary file, and a gene file, this is the gff file created byRastK, as input. The script ""all_blast.py" needs a sequence file, this is the fasta file of the metagenome, a mutation file, this is a tabdelimited variant summary file, and a gene file, this is a txt file with the Blast2GO output, as input. The output for "all.py" is a tabdelimited file with the following columns: Chromosome id, original amino acid, altered amino acid, synonymous/non-synonymousand the gene id. The output for "all_blast.py" is a tab delimited file with the following columns: Chromosome id, original amino acid,altered amino acid, synonymous/non-synonymous, the gene name and the gene id.

The last script that was used was "kegg.py". This script takes the output from the second program and uses this to create afasta file with the protein sequences of the mutated genes. The script needs a sequence file, this is the fasta file of the metagenome, a mutation file, this is the tab delimited file generated by thesecond program. The output is a fasta file with the protein sequences of the mutated genes and a tab delimited file with the gene id,the total number of synonymous mutations and the total number of non-synonymous mutations for that gene.
